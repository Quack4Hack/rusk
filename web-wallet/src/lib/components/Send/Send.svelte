<svelte:options immutable={true} />

<script>
  import { fade } from "svelte/transition";
  import { createEventDispatcher, onMount } from "svelte";
  import {
    mdiAlertOutline,
    mdiArrowUpBoldBoxOutline,
    mdiWalletOutline,
  } from "@mdi/js";
  import { areValidGasSettings, deductLuxFeeFrom } from "$lib/contracts";
  import { duskToLux, luxToDusk } from "$lib/dusk/currency";
  import { validateAddress } from "$lib/dusk/string";
  import { logo } from "$lib/dusk/icons";
  import {
    AnchorButton,
    Badge,
    Button,
    Icon,
    Stepper,
    Textbox,
    Wizard,
    WizardStep,
  } from "$lib/dusk/components";
  import { toast } from "$lib/dusk/components/Toast/store";
  import {
    AppAnchorButton,
    Banner,
    ContractStatusesList,
    GasFee,
    GasSettings,
    OperationResult,
    ScanQR,
  } from "$lib/components";

  /** @type {(to: string, amount: number, gasPrice: number, gasLimit: number) => Promise<string>} */
  export let execute;

  /** @type {(amount: number) => string} */
  export let formatter;

  /** @type {ContractGasSettings} */
  export let gasSettings;

  /** @type {bigint} */
  export let spendable;

  /** @type {ContractStatus[]} */
  export let statuses;

  /** @type {GasStoreContent} */
  export let gasLimits;

  /** @type {boolean} */
  export let enableAllocateButton = false;

  /** @type {boolean} */
  export let enableMoonlightTransactions = false;

  /** @type {number} */
  let amount = 1;

  /** @type {string} */
  let address = "";

  /** @type {import("qr-scanner").default} */
  let scanner;

  /** @type {import("..").ScanQR} */
  let scanQrComponent;

  /** @type {HTMLInputElement | null} */
  let amountInput;

  /** @type {boolean} */
  let isNextButtonDisabled = false;

  /** @type {boolean} */
  let isGasValid = false;

  let { gasLimit, gasPrice } = gasSettings;

  const minAmount = 0.000000001;
  const steps = [
    { label: "Address" },
    { label: "Amount" },
    { label: "Review" },
    { label: "Done" },
  ];
  const dispatch = createEventDispatcher();

  onMount(() => {
    amountInput = document.querySelector(".operation__input-field");
    isGasValid = areValidGasSettings(gasPrice, gasLimit);
  });

  $: luxFee = gasLimit * gasPrice;
  $: fee = formatter(luxToDusk(BigInt(luxFee)));
  $: maxSpendable = deductLuxFeeFrom(luxToDusk(spendable), luxFee);
  $: isAmountValid = amount >= minAmount && amount <= maxSpendable;
  $: totalLuxFee = BigInt(luxFee) + (amount ? duskToLux(amount) : 0n);
  $: isFeeWithinLimit = totalLuxFee <= spendable;
  $: isNextButtonDisabled = !(isAmountValid && isGasValid && isFeeWithinLimit);
  $: addressValidationResult = validateAddress(address);
  $: isMoonlightTransaction = addressValidationResult.type === "account";
  /* eslint-disable no-sequences, no-unused-expressions */
  $: isMoonlightTransaction,
    dispatch("keyChange", {
      type: addressValidationResult.type,
    });
  /* eslint-enable no-sequences, no-unused-expressions */

  function setMaxAmount() {
    if (!isGasValid) {
      toast("error", "Please set valid gas settings first", mdiAlertOutline);
      return;
    }

    if (spendable < BigInt(luxFee)) {
      toast(
        "error",
        "You don't have enough DUSK to cover the transaction fee",
        mdiAlertOutline
      );
      return;
    }

    if (amountInput) {
      amountInput.value = maxSpendable.toString();
    }

    amount = maxSpendable;
  }

  let activeStep = 0;

  /**
   * Validates an address/account depending on moonlight transactions being enabled.
   *
   * Note. This function can be removed when the VITE_FEATURE_MOONLIGHT_TRANSACTIONS flag is removed.
   *
   * @param {{isValid: boolean, type?: "address" | "account"}} validationResult
   */
  function isValid(validationResult) {
    return !validationResult.isValid
      ? true
      : isMoonlightTransaction
        ? !enableMoonlightTransactions
        : false;
  }
</script>

<div class="operation">
  <Wizard steps={steps.length} let:key>
    <div slot="stepper">
      <Stepper {activeStep} {steps} showStepLabelWhenInactive={false} />
    </div>
    <WizardStep
      step={0}
      {key}
      backButton={{
        disabled: false,
        href: "/dashboard",
        isAnchor: true,
      }}
      nextButton={{
        action: () => {
          activeStep = 1;
        },
        disabled: isValid(addressValidationResult),
      }}
    >
      <div in:fade|global class="operation__send">
        <div class="operation__address-wrapper">
          <p>Enter address:</p>
          <Button
            disabled={!scanner}
            size="small"
            on:click={() => {
              scanQrComponent.startScan();
            }}
            text="SCAN QR"
          />
        </div>
        <Textbox
          required
          className={`operation__send-address ${!addressValidationResult.isValid ? "operation__send-address--invalid" : ""}`}
          type="multiline"
          bind:value={address}
        />
        {#if addressValidationResult.type === "account"}
          <Banner
            title="Public account detected"
            variant={enableMoonlightTransactions ? "info" : "warning"}
          >
            {#if enableMoonlightTransactions}
              <p>
                This transaction will be public and sent from your <strong
                  >unshielded</strong
                > account.
              </p>
            {:else}
              <p>Public transactions are currently unavailable.</p>
            {/if}
          </Banner>
        {/if}
        <ContractStatusesList items={statuses}>
          {#if enableAllocateButton}
            <AppAnchorButton
              className="allocate-button"
              href="/dashboard/allocate"
              text="Shield more DUSK"
              variant="tertiary"
            />
          {/if}
        </ContractStatusesList>
        <ScanQR
          bind:this={scanQrComponent}
          bind:scanner
          on:scan={(event) => {
            address = event.detail;
          }}
        />
      </div>
    </WizardStep>
    <WizardStep
      step={1}
      {key}
      backButton={{
        action: () => {
          activeStep = 0;
        },
      }}
      nextButton={{
        action: () => {
          activeStep = 2;
        },
        disabled: isNextButtonDisabled,
      }}
    >
      <div in:fade|global class="operation__send">
        <ContractStatusesList items={statuses}>
          {#if enableAllocateButton}
            <AppAnchorButton
              className="allocate-button"
              href="/dashboard/allocate"
              text="Shield more DUSK"
              variant="tertiary"
            />
          {/if}
        </ContractStatusesList>

        <div class="operation__amount-wrapper">
          <p>Enter amount:</p>
          <Button
            size="small"
            variant="tertiary"
            on:click={setMaxAmount}
            text="USE MAX"
          />
        </div>

        <div class="operation__input-wrapper">
          <Textbox
            className="operation__input-field"
            bind:value={amount}
            required
            type="number"
            min={minAmount}
            max={maxSpendable}
            step="0.000000001"
          />
          <Icon
            data-tooltip-id="main-tooltip"
            data-tooltip-text="DUSK"
            path={logo}
          />
        </div>

        <GasSettings
          {fee}
          limit={gasSettings.gasLimit}
          limitLower={gasLimits.gasLimitLower}
          limitUpper={gasLimits.gasLimitUpper}
          price={gasSettings.gasPrice}
          priceLower={gasLimits.gasPriceLower}
          on:gasSettings={(event) => {
            isGasValid = areValidGasSettings(
              event.detail.price,
              event.detail.limit
            );

            if (isGasValid) {
              gasPrice = event.detail.price;
              gasLimit = event.detail.limit;
            }
          }}
        />
      </div>
    </WizardStep>
    <WizardStep
      step={2}
      {key}
      backButton={{
        action: () => {
          activeStep = 1;
        },
      }}
      nextButton={{
        action: () => {
          activeStep = 3;
        },
        icon: { path: mdiArrowUpBoldBoxOutline, position: "before" },
        label: "SEND",
        variant: "primary",
      }}
    >
      <div in:fade|global class="operation__send">
        <ContractStatusesList items={statuses} />

        <Badge
          className="operation__review-notice"
          text="REVIEW TRANSACTION"
          variant="warning"
        />

        <dl class="operation__review-transaction">
          <dt class="review-transaction__label">
            <Icon path={mdiArrowUpBoldBoxOutline} />
            <span>Amount:</span>
          </dt>
          <dd class="review-transaction__value operation__review-amount">
            <span>{formatter(amount)}</span>
            <Icon
              className="dusk-amount__icon"
              path={logo}
              data-tooltip-id="main-tooltip"
              data-tooltip-text="DUSK"
            />
          </dd>
        </dl>

        <dl class="operation__review-transaction">
          <dt class="review-transaction__label">
            <Icon path={mdiWalletOutline} />
            <span>To:</span>
          </dt>
          <dd class="operation__review-address">
            <span>{address}</span>
          </dd>
        </dl>

        <GasFee {fee} />
      </div>
    </WizardStep>
    <WizardStep step={3} {key} showNavigation={false}>
      <OperationResult
        errorMessage="Transaction failed"
        operation={execute(address, amount, gasPrice, gasLimit)}
        pendingMessage="Processing transaction"
        successMessage="Transaction completed"
      >
        <svelte:fragment slot="success-content" let:result={hash}>
          {#if hash}
            <AnchorButton
              href={`/explorer/transactions/transaction?id=${hash}`}
              text="VIEW ON BLOCK EXPLORER"
              rel="noopener noreferrer"
              target="_blank"
            />
          {/if}
        </svelte:fragment>
      </OperationResult>
    </WizardStep>
  </Wizard>
</div>

<style lang="postcss">
  .operation {
    &__send {
      display: flex;
      flex-direction: column;
      gap: 1.2em;
    }
    &__review-address {
      background-color: transparent;
      border: 1px solid var(--primary-color);
      border-radius: 1.5em;
      padding: 0.75em 1em;
      width: 100%;
      line-break: anywhere;
    }

    &__address-wrapper,
    &__amount-wrapper,
    &__input-wrapper {
      display: flex;
      justify-content: space-between;
      align-items: center;
      width: 100%;
    }

    &__input-wrapper {
      column-gap: var(--default-gap);
    }

    &__review-transaction {
      display: flex;
      flex-direction: column;
      gap: 0.625em;
    }

    &__review-amount {
      justify-content: flex-start;
    }

    :global(&__review-notice) {
      text-align: center;
    }

    :global(.dusk-amount__icon) {
      width: 1em;
      height: 1em;
      flex-shrink: 0;
    }
  }

  .review-transaction__label,
  .review-transaction__value {
    display: inline-flex;
    align-items: center;
    gap: var(--small-gap);
  }

  .review-transaction__value {
    font-weight: bold;
  }

  :global(.dusk-textbox.operation__send-address) {
    resize: vertical;
    min-height: 5em;
    max-height: 10em;
  }

  :global(.dusk-textbox.operation__send-address--invalid) {
    color: var(--error-color);
  }

  :global(.allocate-button) {
    width: 100%;
  }
</style>
