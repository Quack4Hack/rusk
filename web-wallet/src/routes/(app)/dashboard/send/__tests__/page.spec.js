import { afterEach, describe, expect, it, vi } from "vitest";
import { cleanup, render } from "@testing-library/svelte";
import Send from "../+page.svelte";

vi.mock("$lib/dusk/string", async (importOriginal) => {
  /** @type {typeof import("$lib/dusk/string")} */
  const original = await importOriginal();

  return {
    ...original,
    randomUUID: () => "some-generated-id",
  };
});

vi.useFakeTimers();

describe("Send", () => {
  afterEach(cleanup);

  it("should render the send page", async () => {
    const { container } = render(Send);

    await vi.advanceTimersToNextTimerAsync();

    expect(container.firstChild).toMatchSnapshot();
  });
});
