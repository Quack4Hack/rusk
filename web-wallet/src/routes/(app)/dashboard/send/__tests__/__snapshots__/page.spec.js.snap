// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Send > should render the send page 1`] = `
<div
  class="dusk-card dusk-card--gap-medium"
  reverse="true"
>
  <div
    class="dusk-card__header-container"
  >
    <header
      class="dusk-card__header"
      slot="header"
    >
      <div
        class="dusk-card__heading-container"
      >
        <h3
          class="h4"
        >
          Send
        </h3>
         
        <div
          class="dusk-card__heading-icons"
        >
          
          <svg
            class="dusk-icon dusk-icon--size--default"
            role="graphics-symbol"
            viewBox="0 0 24 24"
          >
            <path
              d="M5,17.59L15.59,7H9V5H19V15H17V8.41L6.41,19L5,17.59Z"
            />
          </svg>
          
        </div>
      </div>
    </header>
  </div>
   
  <div
    class="dusk-card__body-container"
  >
    <div
      class="operation svelte-1tivzwh"
    >
      <div
        class="dusk-wizard"
      >
         
         
        <div
          class="operation__send svelte-1tivzwh"
        >
          <dl
            class="contract-statuses svelte-19157rx"
          >
            <dt
              class="contract-statuses__label svelte-19157rx"
            >
              Spendable
               
            </dt>
            <dd
              class="contract-statuses__value svelte-19157rx"
            >
              <span>
                0.000000000
              </span>
               
              <svg
                class="dusk-icon dusk-icon--size--default contract-statuses__icon"
                data-tooltip-id="main-tooltip"
                data-tooltip-text="DUSK"
                role="graphics-symbol"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12.335 0.005a12.186 12.186 0 0 0 -2.58 0.204C4.2 1.259 0 6.14 0 11.999 0 17.859 4.203 22.74 9.759 23.79c0.726 0.138 1.476 0.21 2.241 0.21 6.789 0 12.261 -5.64 11.991 -12.49C23.741 5.249 18.599 0.173 12.335 0.005Zm0.207 21.58c-0.204 0.012 -0.336 -0.222 -0.21 -0.384 1.974 -2.541 3.153 -5.733 3.153 -9.202 0 -3.468 -1.176 -6.663 -3.153 -9.202 -0.126 -0.162 0.003 -0.396 0.207 -0.384 5.052 0.282 9.06 4.465 9.06 9.586 0 5.121 -4.005 9.304 -9.057 9.586Z"
                />
              </svg>
              
               
            </dd>
          </dl>
           
          <div
            class="operation__send-amount operation__space-between svelte-1tivzwh"
          >
            <p>
              Enter amount:
            </p>
             
            <button
              class="dusk-button dusk-button--type--button dusk-button--variant--tertiary dusk-button--size--small"
              type="button"
            >
              <span
                class="dusk-button__text"
              >
                USE MAX
              </span>
            </button>
          </div>
           
          <div
            class="operation__send-amount operation__input svelte-1tivzwh"
          >
            <input
              class="dusk-textbox dusk-textbox-number operation__input-field"
              max="-0.02"
              min="1e-9"
              required=""
              step="0.000000001"
              type="number"
            />
            
             
            <svg
              class="dusk-icon dusk-icon--size--default"
              data-tooltip-id="main-tooltip"
              data-tooltip-text="DUSK"
              role="graphics-symbol"
              viewBox="0 0 24 24"
            >
              <path
                d="M12.335 0.005a12.186 12.186 0 0 0 -2.58 0.204C4.2 1.259 0 6.14 0 11.999 0 17.859 4.203 22.74 9.759 23.79c0.726 0.138 1.476 0.21 2.241 0.21 6.789 0 12.261 -5.64 11.991 -12.49C23.741 5.249 18.599 0.173 12.335 0.005Zm0.207 21.58c-0.204 0.012 -0.336 -0.222 -0.21 -0.384 1.974 -2.541 3.153 -5.733 3.153 -9.202 0 -3.468 -1.176 -6.663 -3.153 -9.202 -0.126 -0.162 0.003 -0.396 0.207 -0.384 5.052 0.282 9.06 4.465 9.06 9.586 0 5.121 -4.005 9.304 -9.057 9.586Z"
              />
            </svg>
            
          </div>
           
          <div
            class="gas-settings svelte-1daay07"
          >
            <dl
              class="gas-settings__edit svelte-1daay07"
            >
              <div
                class="gas-fee svelte-1c6b481"
              >
                <span>
                  Fee:
                </span>
                 
                <div
                  class="gas-fee__amount svelte-1c6b481"
                >
                  <div
                    class="gas-fee__amount-value svelte-1c6b481"
                  >
                    <span
                      class="svelte-1c6b481"
                    >
                      max
                    </span>
                     
                    <span
                      class="svelte-1c6b481"
                    >
                      0.020000000
                    </span>
                  </div>
                   
                  <svg
                    class="dusk-icon dusk-icon--size--default dusk-amount__icon"
                    data-tooltip-id="main-tooltip"
                    data-tooltip-text="DUSK"
                    role="graphics-symbol"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12.335 0.005a12.186 12.186 0 0 0 -2.58 0.204C4.2 1.259 0 6.14 0 11.999 0 17.859 4.203 22.74 9.759 23.79c0.726 0.138 1.476 0.21 2.241 0.21 6.789 0 12.261 -5.64 11.991 -12.49C23.741 5.249 18.599 0.173 12.335 0.005Zm0.207 21.58c-0.204 0.012 -0.336 -0.222 -0.21 -0.384 1.974 -2.541 3.153 -5.733 3.153 -9.202 0 -3.468 -1.176 -6.663 -3.153 -9.202 -0.126 -0.162 0.003 -0.396 0.207 -0.384 5.052 0.282 9.06 4.465 9.06 9.586 0 5.121 -4.005 9.304 -9.057 9.586Z"
                    />
                  </svg>
                  
                </div>
              </div>
              <dd>
                <button
                  class="dusk-button dusk-button--type--button dusk-button--variant--tertiary dusk-button--size--small"
                  type="button"
                >
                  <span
                    class="dusk-button__text"
                  >
                    EDIT
                  </span>
                </button>
              </dd>
            </dl>
             
          </div>
        </div>
         
        <div
          class="dusk-wizard__step-navigation"
        >
          <a
            aria-disabled="false"
            class="dusk-anchor dusk-anchor--on-surface dusk-anchor-button dusk-anchor-button--variant--tertiary dusk-anchor-button--size--default dusk-icon-button--labeled"
            href="/some-base-path/dashboard"
          >
            <svg
              class="dusk-icon dusk-icon--size--default dusk-anchor-button__icon"
              role="graphics-symbol"
              viewBox="0 0 24 24"
            >
              <path
                d="M20,11V13H8L13.5,18.5L12.08,19.92L4.16,12L12.08,4.08L13.5,5.5L8,11H20Z"
              />
            </svg>
            
             
            <span
              class="dusk-anchor-button__text"
            >
              Back
            </span>
            
            
          </a>
          
           
          <button
            class="dusk-button dusk-button--type--button dusk-button--variant--tertiary dusk-button--size--default dusk-icon-button--labeled"
            disabled=""
            type="button"
          >
            <span
              class="dusk-button__text"
            >
              Next
            </span>
             
            <svg
              class="dusk-icon dusk-icon--size--default dusk-button__icon"
              role="graphics-symbol"
              viewBox="0 0 24 24"
            >
              <path
                d="M4,11V13H16L10.5,18.5L11.92,19.92L19.84,12L11.92,4.08L10.5,5.5L16,11H4Z"
              />
            </svg>
            
          </button>
        </div>
        
        
         
        
         
        
         
        
      </div>
    </div>
  </div>
   
</div>
`;
